### PR TITLE
batch api HF adapter for ring-flash-attn; cleanup and improvements

### PR DESCRIPTION
**Pull request by djsaunde**
# Description

Added an adapter similar to [`ring-flash-attn`'s](https://github.com/zhuzilin/ring-flash-attention/blob/c587a0def9eaac7709e0d439757033178f30069f/ring_flash_attn/adapters/hf_adapter.py#L93) for the ring attention batch API.

We should probably upstream this adapter at some point.

## Motivation and Context

This allows us to support sequence parallelism for the non-sample packed case, and allows for microbatch size > 1 (in that case). This will be useful for e.g. GRPO training, where we're currently not sample packing and use >1 microbatch size.

## How has this been tested?

Varied microbatch_size in [1, 2, 4] and 

Added smoke tests for these added features.

## Screenshots (if appropriate)

Example with varying SP in 1, 2, 4 and fixing microbatch size at 2:

W B Chart 4_10_2025, 11_59_51 PM

FYI: losses ~match due to a fix in the logged loss calculation in the huggingface trainer (we may upstream).

## TODO / follow-ups

- [ ] Gradient explosion with `pad_to_sequence_len: false`
- [x] Additional HF adapters for the remaining `ring-flash-attn` batch functions (`zigzag` and `stripe`)
- [ ] Additional HF adapters for the remaining varlen functions (`ring`, `zigzag`)
- [ ] Move to `https://github.com/feifeibear/long-context-attention` instead?
  - This might just entail adding HF adapters for these that match what we're currently using + testing for perf gains / different scenarios

### CODE REVIEW

**Comment by djsaunde** [comment]
SP = 4, microbatch size = 2, batch ring vs. batch stripe vs. batch zigzag ~match. There are some notable diffs for zigzag on a few steps, however 🤔

W B Chart 4_11_2025, 12_47_58 AM (1)

[wandb](https://wandb.ai/axolotl-ai/seq-parallel-batch?nw=nwuserdanaxolotl&panelDisplayName=train%2Floss&panelSectionName=train)

File: docs/sequence_parallelism.qmd [hunk 1]
```diff
@@ -27,6 +27,9 @@ To enable sequence parallelism, add the following to your configuration file:
 sequence_parallel_degree: 4  # Split sequences across 4 GPUs
 # Optional; strides across the key dimension. Larger values use more memory but should make training faster.
 heads_k_stride: 1
+# Optional; one of "varlen_llama3", "batch_ring", "batch_zigzag", "batch_stripe". Defaults to
+"varlen_llama3" when `sample_packing: true`, and "batch_ring" otherwise.
```
Comment (salmanmohammadi) [inline docs/sequence_parallelism.qmd]:
```suggestion
# "varlen_llama3" when `sample_packing: true`, and "batch_ring" otherwise.
```

File: src/axolotl/monkeypatch/attention/ring_attn/patch.py [hunk 2]
```diff
@@ -40,22 +43,58 @@ def set_ring_attn_group(ring_attn_group: dist.ProcessGroup | None):
     RING_ATTN_GROUP = ring_attn_group
 
 
-def register_ring_attn(sequence_parallel_degree: int, heads_k_stride: int | None):
+class RingAttnFunc(str, Enum):
+    """Enum class for supported `ring-flash-attn` implementations"""
+
+    # VARLEN_RING = "varlen_ring"
+    # VARLEN_ZIGZAG = "varlen_zigzag"
```
Comment (SalmanMohammadi) [inline src/axolotl/monkeypatch/attention/ring_attn/patch.py:50]:
```suggestion
```
nit

File: src/axolotl/monkeypatch/attention/ring_attn/patch.py [hunk 3]
```diff
@@ -40,22 +43,58 @@ def set_ring_attn_group(ring_attn_group: dist.ProcessGroup | None):
     RING_ATTN_GROUP = ring_attn_group
 
 
-def register_ring_attn(sequence_parallel_degree: int, heads_k_stride: int | None):
+class RingAttnFunc(str, Enum):
+    """Enum class for supported `ring-flash-attn` implementations"""
+
+    # VARLEN_RING = "varlen_ring"
```
Comment (salmanmohammadi) [inline src/axolotl/monkeypatch/attention/ring_attn/patch.py:49]: how come these are commented out?

File: src/axolotl/monkeypatch/attention/ring_attn/patch.py [hunk 4]
```diff
@@ -40,22 +43,58 @@ def set_ring_attn_group(ring_attn_group: dist.ProcessGroup | None):
...
+    BATCH_RING = "batch_ring"
+    BATCH_ZIGZAG = "batch_zigzag"
+    BATCH_STRIPE = "batch_stripe"
+
+
+def register_ring_attn(
+    sequence_parallel_degree: int,
+    heads_k_stride: int | None,
+    sample_packing: bool,
+    ring_attn_func: str | None,
```
Comment (salmanmohammadi) [inline src/axolotl/monkeypatch/attention/ring_attn/patch.py]: nit: could type with `RingAttnFunc`? Or just indicate in the docstring it should be one of those values.

File: src/axolotl/utils/collators/batching.py [hunk 5]
```diff
@@ -157,18 +157,20 @@ def apply_sequence_parallelism(
...
+            update_ring_attn_params(
+                input_ids=batch["input_ids"], position_ids=batch.get("position_ids")
+            )
 
         # Slice batch for sequence parallel processing
-        keys_to_slice = ["input_ids", "attention_mask", "labels", "position_ids"]
-        for key in keys_to_slice:
-            if key in batch:
+        for key in batch:
+            if batch[key].size(1) == total_seq_len:
```
Comment (salmanmohammadi) [inline src/axolotl/utils/collators/batching.py]: Will there ever be keys in the batch that fulfill ` batch[key].size(1) == total_seq_len` but which we don't want to slice?

File: tests/e2e/multigpu/test_sp.py [hunk 6]
```diff
@@ -17,8 +17,14 @@
 class TestSequenceParallelism:
     """Test case for training with sequence parallelism enabled"""
 
-    def test_sequence_parallel_training(self, temp_dir):
-        # pylint: disable=duplicate-code
+    def _run_sequence_parallel_test(
```
Comment (salmanmohammadi) [inline tests/e2e/multigpu/test_sp.py:21]: Worth throwing a `parametrize` here to exercise the different options for this test?

**Review by salmanmohammadi — APPROVED** [review]

File: src/axolotl/monkeypatch/attention/ring_attn/patch.py (hunk 3 above)
Comment (djsaunde) [reply src/axolotl/monkeypatch/attention/ring_attn/patch.py:49]: not yet supported

File: src/axolotl/utils/collators/batching.py (hunk 5 above)
Comment (djsaunde) [reply src/axolotl/utils/collators/batching.py]: I don't think so? The thought behind this was to extend it to arbitrary sequence data.

File: src/axolotl/monkeypatch/attention/ring_attn/patch.py (hunk 3 above)
Comment (djsaunde) [reply src/axolotl/monkeypatch/attention/ring_attn/patch.py:49]: would need another adapter module, but would be pretty straightforward to add to this PR

File: tests/e2e/multigpu/test_sp.py (hunk 6 above)
Comment (djsaunde) [reply tests/e2e/multigpu/test_sp.py:21]: Maybe, though I don't want to add too many tests + extend the runtime too much.

File: src/axolotl/monkeypatch/attention/ring_attn/patch.py (hunk 4 above)
Comment (winglian) [reply src/axolotl/monkeypatch/attention/ring_attn/patch.py]: Yeah. Enum or Literal type
